File: svg_mobile_app/svg_mobile_app/report/project_expenses_report/project_expenses_report.js
// Copyright (c) 2025, SVG and contributors
// For license information, please see license.txt

frappe.query_reports["Project Expenses Report"] = {
    "filters": [
        {
            "fieldname": "project_contractor",
            "label": __("Project Contractor"),
            "fieldtype": "Link",
            "options": "Project Contractors",
            "width": "200px"
        },
        {
            "fieldname": "employee",
            "label": __("Employee"),
            "fieldtype": "Link",
            "options": "Employee",
            "width": "200px"
        },
        {
            "fieldname": "expense_type",
            "label": __("Expense Type"),
            "fieldtype": "Link",
            "options": "Expense Claim Type",
            "width": "200px"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date (Expense)"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "width": "200px"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date (Expense)"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "width": "200px"
        },
        {
            "fieldname": "posting_from_date",
            "label": __("From Date (Posting)"),
            "fieldtype": "Date",
            "width": "200px"
        },
        {
            "fieldname": "posting_to_date",
            "label": __("To Date (Posting)"),
            "fieldtype": "Date",
            "width": "200px"
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "amount" && data && data.amount) {
            value = `<div style="text-align: right; font-weight: bold; color: #2e7d32;">${value}</div>`;
        }
        
        if (column.fieldname == "project_contractor" && data && data.project_contractor) {
            value = `<a href="/app/project-contractors/${data.project_contractor}" target="_blank">${value}</a>`;
        }
        
        if (column.fieldname == "expense_claim" && data && data.expense_claim) {
            value = `<a href="/app/expense-claim/${data.expense_claim}" target="_blank">${value}</a>`;
        }
        
        return value;
    }
};

File: svg_mobile_app/svg_mobile_app/report/project_expenses_report/project_expenses_report.py
# Copyright (c) 2025, SVG and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    """
    Main execution function for the Project Expenses Report
    Returns columns and data for the report
    """
    if not filters:
        filters = {}
    
    # Define columns
    columns = [
        {"label": _("Expense Date"), "fieldname": "expense_date", "fieldtype": "Date", "width": 100},
        {"label": _("Expense Type"), "fieldname": "expense_type", "fieldtype": "Link", "options": "Expense Claim Type", "width": 120},
        {"label": _("Description"), "fieldname": "description", "fieldtype": "Data", "width": 200},
        {"label": _("Amount"), "fieldname": "amount", "fieldtype": "Currency", "width": 100},
        {"label": _("Employee"), "fieldname": "employee_name", "fieldtype": "Data", "width": 150},
        {"label": _("Employee ID"), "fieldname": "employee", "fieldtype": "Link", "options": "Employee", "width": 100},
        {"label": _("Expense Claim"), "fieldname": "expense_claim", "fieldtype": "Link", "options": "Expense Claim", "width": 120},
        {"label": _("Project Contractor"), "fieldname": "project_contractor", "fieldtype": "Link", "options": "Project Contractors", "width": 150},
        {"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 100}
    ]
    
    # Get values from filters
    project_contractor = filters.get("project_contractor")
    employee = filters.get("employee")
    expense_type = filters.get("expense_type")
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    posting_from_date = filters.get("posting_from_date")
    posting_to_date = filters.get("posting_to_date")
    
    # Build conditions
    conditions = []
    params = {}
    
    # Base conditions
    conditions.append("`tabExpense Claim`.`docstatus` = 1")
    conditions.append("`tabExpense Claim`.`approval_status` = 'Approved'")
    
    # Filter conditions
    if project_contractor:
        conditions.append("`tabExpense Claim Detail`.`for_project` = %(project_contractor)s")
        params["project_contractor"] = project_contractor
    if employee:
        conditions.append("`tabExpense Claim`.`employee` = %(employee)s")
        params["employee"] = employee
    if expense_type:
        conditions.append("`tabExpense Claim Detail`.`expense_type` = %(expense_type)s")
        params["expense_type"] = expense_type
    if from_date:
        conditions.append("`tabExpense Claim Detail`.`expense_date` >= %(from_date)s")
        params["from_date"] = from_date
    if to_date:
        conditions.append("`tabExpense Claim Detail`.`expense_date` <= %(to_date)s")
        params["to_date"] = to_date
    if posting_from_date:
        conditions.append("`tabExpense Claim`.`posting_date` >= %(posting_from_date)s")
        params["posting_from_date"] = posting_from_date
    if posting_to_date:
        conditions.append("`tabExpense Claim`.`posting_date` <= %(posting_to_date)s")
        params["posting_to_date"] = posting_to_date
    
    # Combine conditions
    conditions_str = " AND ".join(conditions) if conditions else "1=1"
    
    # SQL query - Using for_project field from Expense Claim Detail
    sql = f"""
        SELECT
            `tabExpense Claim Detail`.`expense_date`,
            `tabExpense Claim Detail`.`expense_type`,
            `tabExpense Claim Detail`.`description`,
            `tabExpense Claim Detail`.`amount`,
            `tabExpense Claim`.`employee_name`,
            `tabExpense Claim`.`employee`,
            `tabExpense Claim`.`name` AS `expense_claim`,
            `tabExpense Claim Detail`.`for_project` AS `project_contractor`,
            `tabExpense Claim`.`posting_date`
        FROM
            `tabExpense Claim Detail`
        INNER JOIN
            `tabExpense Claim` ON `tabExpense Claim Detail`.`parent` = `tabExpense Claim`.`name`
        WHERE
            {conditions_str}
        ORDER BY
            `tabExpense Claim Detail`.`expense_date` DESC
    """
    
    # Execute query
    data = frappe.db.sql(sql, params, as_dict=True)
    
    return columns, data
